### PR TITLE
support Internet Explorer User Proxy

### PR DESCRIPTION
**Pull request by sclassen**
PR with the code from zip file attached to #385

### CODE REVIEW

File: netx/net/sourceforge/jnlp/runtime/JNLPProxySelector.java [hunk 1]
```diff
@@ -48,6 +56,18 @@
     public static final int PROXY_TYPE_MANUAL = 1;
     public static final int PROXY_TYPE_AUTO = 2;
     public static final int PROXY_TYPE_BROWSER = 3;
+    public static final String PROXY_REGISTRY_KEY =
+            "Software\\Microsoft\\Windows\\CurrentVersion\\Internet Settings";
+    public static final String PROXY_DWORD_REGISTRY_KEY =
+            "Software\\Microsoft\\Windows\\CurrentVersion\\Internet Settings";
+    //    public static final String PROXY_ENABLE_REGISTRY_VAL =
+//             "ProxyEnable";
```
Comment (judovana) [inline netx/net/sourceforge/jnlp/runtime/JNLPProxySelector.java]: indentation

File: netx/net/sourceforge/jnlp/runtime/JNLPProxySelector.java [hunk 2]
```diff
@@ -180,36 +199,95 @@ public void connectFailed(URI uri, SocketAddress sa, IOException ioe) {
...
     @Override
     public List<Proxy> select(URI uri) {
-        OutputController.getLogger().log("Selecting proxy for: " + uri);
-        
-        if (inBypassList(uri)) {
-            List<Proxy> proxies = Arrays.asList(new Proxy[] { Proxy.NO_PROXY });
-            OutputController.getLogger().log("Selected proxies: " + Arrays.toString(proxies.toArray()));
-            return proxies;
-        }
+//        OutputController.getLogger().log("Selecting proxy for: " + uri);
```
Comment (judovana) [inline netx/net/sourceforge/jnlp/runtime/JNLPProxySelector.java]: drop thsi comment

File: netx/net/sourceforge/jnlp/runtime/JNLPProxySelector.java [hunk 3]
```diff
@@ -180,36 +199,95 @@ public void connectFailed(URI uri, SocketAddress sa, IOException ioe) {
...
+        OutputController.getLogger().log(OutputController.Level.MESSAGE_DEBUG, "Selecting proxy for: " + uri);
+
+
+
+
+//        if (inBypassList(uri)) {
+//            List<Proxy> proxies = Arrays.asList(new Proxy[] { Proxy.NO_PROXY });
+//            OutputController.getLogger().log(OutputController.Level.MESSAGE_DEBUG, "Selected proxies: " + Arrays.toString(proxies.toArray()));
+//            return proxies;
+//        }
```
Comment (judovana) [inline netx/net/sourceforge/jnlp/runtime/JNLPProxySelector.java]: Should baypass list really be dropped?

File: netx/net/sourceforge/jnlp/runtime/JNLPProxySelector.java [hunk 4]
```diff
@@ -180,36 +199,95 @@ public void connectFailed(URI uri, SocketAddress sa, IOException ioe) {
...
+                    }
+                    proxyOver = WinRegistry.readString(
+                            WinRegistry.HKEY_CURRENT_USER,
+                            PROXY_REGISTRY_KEY,
+                            PROXY_SERVER_OVERRIDE_VAL, 0);
+                    autoConfigString = WinRegistry.readString(
+                            WinRegistry.HKEY_CURRENT_USER,
+                            PROXY_REGISTRY_KEY,
+                            PROXY_AUTOCONFIG_URL_VAL, 0);
+                    if (proxyServer.equalsIgnoreCase("localhost:1234")) {
```
Comment (judovana) [inline netx/net/sourceforge/jnlp/runtime/JNLPProxySelector.java]: What kind of convention is this?

File: netx/net/sourceforge/jnlp/runtime/JNLPProxySelector.java [hunk 5]
```diff
@@ -180,36 +199,95 @@ public void connectFailed(URI uri, SocketAddress sa, IOException ioe) {
...
+                        OutputController.getLogger().log(OutputController.Level.MESSAGE_DEBUG, "IE has wrong ProxyServer Registry Key Value: "  + proxyServer + ", Java uses DIRECT connection");
+                        proxies.add(Proxy.NO_PROXY);
+                    }
+                    else if (proxyServer != null && proxyOver !=null) {
+                        OutputController.getLogger().log(OutputController.Level.MESSAGE_DEBUG, "IE has ProxyServer & ProxyOverride Registry Key and Proxy Server value is: " + proxyServer);
+                        proxies.addAll(getFromWindowsRegistry(uri));
+                    }
+                    else if (autoConfigString != null) {
+                        OutputController.getLogger().log(OutputController.Level.MESSAGE_DEBUG, "IE has PAC File URL Registry Key");
+                        proxies.addAll(getFromPAC(uri));
```
Comment (judovana) [inline netx/net/sourceforge/jnlp/runtime/JNLPProxySelector.java]: Is rhino stioll around to read pac?

File: netx/net/sourceforge/jnlp/runtime/JNLPProxySelector.java [hunk 6]
```diff
@@ -393,6 +471,115 @@ private boolean isLocalHost(String host) {
...
+
+    } // getFromWindowsRegistry
+
+    public static Boolean checkProxyOverride(String proxyOverrides, URI uri) {
+
+        Boolean overrideFound = false;
+        if (proxyOverrides != null) {
+            String[] overrides = proxyOverrides.split(";");
+            for (String override : overrides) {
+//            OutputController.getLogger().log("Proxy override check: Comparing '" +
```
Comment (judovana) [inline netx/net/sourceforge/jnlp/runtime/JNLPProxySelector.java]: I'm in favour to drop all those comment, or to enable it. Logging to most verbose log is good thing to do.

**Comment by sclassen** [comment]
@patris70 can you have a look at the comments above

File: netx/net/sourceforge/jnlp/runtime/JNLPProxySelector.java (hunk 4 above)
Comment (patris70) [reply netx/net/sourceforge/jnlp/runtime/JNLPProxySelector.java]: @judovana

this ist fake Proxy in our environment, I must check this value (localhost:1234).
you can drop it.

**Comment by patris70** [comment]
@judovana 

without WinRegistry, you can't parsing Registry Keys.
With this code, you can parsing DWORD Value (ProxyEnabled) with Reg.exe, but i can't parsing DWORD, because normal User in our environment can't run reg.exe or cmd.exe, because of this, i must using just REG_SZ Values to compare values.

File: netx/net/sourceforge/jnlp/runtime/JNLPProxySelector.java (hunk 3 above)
Comment (patris70) [reply netx/net/sourceforge/jnlp/runtime/JNLPProxySelector.java]: No

File: netx/net/sourceforge/jnlp/runtime/JNLPProxySelector.java (hunk 5 above)
Comment (patris70) [reply netx/net/sourceforge/jnlp/runtime/JNLPProxySelector.java]: yes

**Comment by lherschi** [comment]
I believe with this change, the firefox on linux no longer works.

**Comment by judovana** [comment]
> @judovana
> 
> without WinRegistry, you can't parsing Registry Keys.
> With this code, you can parsing DWORD Value (ProxyEnabled) with Reg.exe too, but i can't parsing DWORD, because normal User in our environment can't run reg.exe or cmd.exe, because of this, i must using just REG_SZ Values to compare values.
> 
> Why isWindows, this code is **just for Windows OS** (worked with 7, 8, 10 clients and 2008, 2012, 2012R2 servers).
> Hope help you

You got mw wrong - yoou had broken linux.  Your code needs a lot of  if (isWindows()) {then windows registry stuff} tuning. Can you pelase elaborate?

The class WinRegistry is of course fundamental, and have to be pushed, butI was not reviewing it.  The responsibility remains on small windows community.

What troubles me indded is the integration of this class with rest of ITW(read linux+mac+firefox and spol) . We are ok to test it for you, but some basic debugging have to start on your end.

In all cases I'm huge fan of this pathc. Thank you fro doing it!

**Comment by Archimedes5000** [comment]
@judova, do you see any chance to integrate this into ITW 2.0 ? What is to do:
- Select (isWindows())
- Removing reg query for reading Registry, see Patris70 last comment. reg query does not work for a User on Terminalserver because they are not allowed to start this exe. Maybe Powershell works.
How can we help ?

**Comment by judovana** [comment]
@Archimedes5000  I have no doubts it will reach 2.0. But somebody have to tune the patch. And that will not be me. I can not fix windows patch. I can just check it did nto broke linux. Sorry.

**Comment by patris70** [comment]
@Archimedes5000 
i told in "our" TS farm, we have prevent that user run cmd or reg.exe.
As info, with this JNLPProySelector.java, you donot need usig reg.exe or PowerShell.

**Comment by judovana** [comment]
Sure. but the patch needs tunning.

**Comment by falmx** [comment]
@Archimedes5000 like your idea very much. Just one question: is there still work going on to solve the issues / comments raised in the discussion above?

**Comment by Archimedes5000** [comment]
@falmx At this time, we have a workaround in our production. The user logonscript reads the registry and creates a pac file. We use powershell in logon process, not reg.exe. Works great for us. 
At this time I do not really see where the jouney is going. 
Icedtea-web is an extension for end users. Most End Users use Windows, not Linux. 
judova told us that we broke linux. For our environment it is ok. We have many Linux Systems in production, but not one linux based Client for the users, if I do not consider thin clients. Our deployment startet with Icedtea-Web 1.7.1 in 2018, production in 04 2019. In 08 2019 we changed immediately to 1.8.3 because of 3 existing cert messages for ITW. 
The new structure in the source 2.0 code currently prevents our extensions which are absolutely necessary for us in production, many were rejected by the project or will be implemented in 2.0. We can not support 2.0 at the moment because the documentation is currently not sufficient or not available at all. Compiling for windows does not really work. See all actions from Patris70 in this project. If this project goes to open Webstart, which Java Version is still supported ? We are currently planning to replace Java 8 at the end of 2023.
My biggest concern at the moment is how can we integrate future security patches into the version ITW 1.8 we need?
Our next step: Waiting for ITW2.0 documentation for windows simliar to ojdkbuild wiki.

Greetings

**Comment by falmx** [comment]
@Archimedes5000 Okay then. I'll have a look at the code in the next weeks and try to solve the points raised. E.g. the "do not break" Linux part.

**Comment by patris70** [comment]
Hi,

here ist PowerShell script
https://github.com/AdoptOpenJDK/IcedTea-Web/issues/152

But at this time at the moment I try to intgrate PowerShell code in JNLPProxySelector.java file.

**Comment by hendrikebbers** [comment]
closed since we will include https://github.com/AdoptOpenJDK/IcedTea-Web/pull/523

**Pull request closed without merging**